File: fuzz/fuzz_targets/diff_base32.py
import json
import os
import sys
from argparse import ArgumentParser

script_dir = os.path.dirname(os.path.abspath(__file__))
bbqr_path = os.path.abspath(os.path.join(script_dir, "..", "..", "BBQr", "python"))
sys.path.insert(0, bbqr_path)
krux_path = os.path.abspath(os.path.join(script_dir, "..", "..", "selfcustody", "src"))
sys.path.insert(0, krux_path)


from krux.bbqr import base32_decode_stream, base32_encode_stream

from bbqr.utils import decode_data, encode_data


def main():
    parser = ArgumentParser(prog="diff_base32")
    parser.add_argument("-d", "--data", type=str)
    args = parser.parse_args()

    data = json.loads(args.data)
    module = data["module"]
    action = data["action"]
    padding = data["padding"]
    input_data = bytes(data["input"])

    if module == "krux":
        if action == "encode":
            encoded = "".join(base32_encode_stream(input_data, add_padding=padding))
            print(json.dumps({"result": encoded}))
        elif action == "decode":
            decoded = base32_decode_stream(input_data.decode("utf-8"))
            print(json.dumps({"result": list(decoded)}))
        else:
            raise ValueError(f"Invalid module '{module}' for module krux")

    elif module == "bbqr":
        if action == "encode":
            encoded = encode_data(input_data)
            print(json.dumps({"result": encoded[1]}))
        else:
            raise ValueError(f"Invalid action '{action}' for module 'bbqr'")

    else:
        raise ValueError(f"Invalid module '{module}'")


if __name__ == "__main__":
    main()

File: fuzz/fuzz_targets/diff_base32.rs
#![no_main]
use base32;
use call_python_script;
use libfuzzer_sys::fuzz_target;

fuzz_target!(|data: &[u8]| {
    for &padding in &[false, true] {
        let alphabet = base32::Alphabet::Rfc4648 { padding };

        let rust_encoded = base32::encode(alphabet, data);
        let krux_encoded: String = call_python_script::run(
            "fuzz/fuzz_targets/diff_base32.py",
            &serde_json::json!({
                "module": "krux",
                "action": "encode",
                "padding": padding,
                "input": data,
            }),
        )
        .expect("krux encode failed");
        assert_eq!(rust_encoded, krux_encoded);

        // BBQr do not have code for padding in utils module
        if !padding {
            let bbqr_encoded: String = call_python_script::run(
                "fuzz/fuzz_targets/diff_base32.py",
                &serde_json::json!({
                    "module": "bbqr",
                    "action": "encode",
                    "padding": padding,
                    "input": data,
                }),
            )
            .expect("bbqr encode failed");

            assert_eq!(krux_encoded, bbqr_encoded);
            assert_eq!(rust_encoded, bbqr_encoded);
        }

        let krux_decoded: Vec<u8> = call_python_script::run(
            "fuzz/fuzz_targets/diff_base32.py",
            &serde_json::json!({
                "module": "krux",
                "action": "decode",
                "padding": padding,
                "input": rust_encoded.as_bytes(),
            }),
        )
        .expect("krux decode failed");

        assert_eq!(data, krux_decoded);
    }
});
